File: ueb6/src/OT.hs
{-# LANGUAGE MultiParamTypeClasses, FlexibleInstances, ExistentialQuantification #-}

module OT where
import Control.Applicative

class Operation o where
  compose :: o -> o -> o
  transform :: o -> o -> (o,o)

class Operation o => Document d o where
  noop :: d -> o
  applyOp :: d -> o -> d

---

data TextAction = Retain Int
                | Insert String
                | Delete Int deriving Show

newtype TextOperation = TextOperation [TextAction] deriving Show

instance Operation TextOperation where
  compose (TextOperation a) (TextOperation b) = TextOperation $ compose' a b where
    compose' [] [] = []
    compose' (Delete n: as) bs   = Delete n : compose' as bs
    compose' as (Insert cs : bs) = Insert cs : compose' as bs
    compose' (Retain n : as) (Retain n' : bs) 
      | n > n'  = Retain n' : compose' (Retain (n-n') : as) bs
      | n < n'  = Retain n : compose' as (Retain (n'-n) : bs)
      | otherwise = Retain n : compose' as bs
    compose' (Retain n : as) (Delete n' : bs) 
      | n > n'  = Delete n' : compose' (Retain (n-n') : as) bs
      | n < n'  = Delete n : compose' as (Retain (n'-n) : bs)
      | otherwise = Delete n : compose' as bs
    compose' (Insert cs : as) (Retain n : bs) 
      | n > length cs = Insert cs : compose' as (Retain (n-length cs) : bs)
      | n < length cs = Insert (take n cs) : compose' (Insert (drop n cs) : as) bs
      | otherwise     = Insert cs : compose' as bs
    compose' (Insert cs : as) (Delete n :bs)  
      | n > length cs = compose' as $ Delete (n - length cs) : bs 
      | n < length cs = compose' ((Insert $ drop n cs) : as) bs
      | otherwise     = compose' as bs
  transform (TextOperation a) (TextOperation b) = pair $ transform' a b where
    pair (as',bs') = (TextOperation as', TextOperation bs')
    transform' [] [] = ([],[])
    transform' (Insert cs : as) bs = case transform' as bs of (as',bs') -> (Insert cs : as', Retain (length cs) : bs')
    transform' as (Insert cs : bs) = case transform' as bs of (as',bs') -> (Retain (length cs) : as', Insert cs : bs')
    transform' (Retain n: as) (Retain n': bs) 
      | n > n' = case transform' (Retain (n-n') : as) bs of (as',bs') -> (Retain n' : as', Retain n' : bs')
      | n < n' = case transform' as (Retain (n'-n) : bs) of (as',bs') -> (Retain n : as', Retain n : bs')  
      | otherwise = case transform' as bs of (as',bs') -> (Retain n: as', Retain n': bs')
    transform' (Delete n : as) (Delete n' : bs)
      | n > n' = transform' (Delete (n-n') : as) bs
      | n < n' = transform' as $ Delete (n'-n) : bs 
      | otherwise = transform' as bs
    transform' (Retain n : as) (Delete n' : bs) 
      | n > n' = case transform' (Retain (n-n') : as) bs of (as',bs') -> (as', Delete n' : bs') 
      | n < n' = case transform' as (Delete (n'-n) : bs) of (as',bs') -> (as', Delete n : bs') 
      | otherwise = case transform' as bs of (as',bs') -> (as', Delete n : bs')
    transform' (Delete n : as) (Retain n' : bs) 
      | n > n' = case transform' (Delete (n-n') : as) bs of (as',bs') -> (Delete n' : as', bs')
      | n < n' = case transform' as (Retain (n'-n) : bs) of (as',bs') -> (Delete n : as', bs')
      | otherwise = case transform' as bs of (as',bs') -> (Delete n: as', bs')

instance Document String TextOperation where
  noop d 
    | length d == 0 = TextOperation [] 
    | otherwise = TextOperation [Retain $ length d]
  applyOp d (TextOperation as) = apply' as d where
    apply' [] [] = []
    apply' (Retain 0 : as) cs = apply' as cs
    apply' (Retain n : as) (c:cs) = c : apply' (Retain (n-1) : as) cs
    apply' (Insert cs : as) d = cs ++ apply' as d
    apply' (Delete 0 : as) cs = apply' as cs
    apply' (Delete n : as) (_:cs) = apply' (Delete (n-1) : as) cs

---

data Client o d = (Operation o, Document d o) => Client {
  clientRevision :: Integer,
  pending :: Maybe o,
  buffer :: Maybe o
}

newClient :: (Operation o, Document d o) => Client o d
newClient = Client 0 Nothing Nothing

localEdit :: Client o d -> o -> (Bool, Client o d)
localEdit (Client rev pending buffer) operation = (null pending, Client rev newPending newBuffer) where
    newPending = pending <|> (Just operation)
    newBuffer  = pending *> fmap (`compose` operation) buffer

remoteEdit :: Client o d -> o -> (o, Client o d)
remoteEdit (Client rev pending buffer) o = case (pending,buffer) of
    (Nothing,Nothing) -> (o, Client (rev + 1) Nothing Nothing)
    (Just p,Nothing)  -> let (p',o') = transform p o in
                             (o',Client (rev + 1) (Just p') Nothing)
    (Just p,Just b)   -> let (p',o') = transform p o
                             (b',o'') = transform b o' in
                             (o'',Client (rev + 1) (Just o') (Just b'))

ack :: Client o d -> (Maybe o, Client o d)
ack (Client rev pending buffer) = (buffer, Client (rev + 1) buffer Nothing)

isSynchronized :: Client o d -> Bool
isSynchronized (Client rev pending buffer) = null pending

---

data Server o d = (Operation o, Document d o) => Server {
    initialState :: d,
    history :: [o]
}


type Cursor = Int 

transformCursor :: TextOperation -> Cursor -> Cursor
transformCursor (TextOperation op) c = foldr transf c $ split op c where 
  transf op c = case op of
    Retain n -> c 
    Insert str -> c + length str
    Delete n -> c - n
  split [] _ = []
  split (x:xs) c 
    | c <= 0 = [] 
    | otherwise = case x of
        Retain n -> 
          if n <= c 
          then x : split xs (c-n) 
          else [Retain c]
        Delete n ->
          if n <= c
          then x : split xs (c-n)
          else [Delete c]
        Insert cs -> 
          if length cs <= c 
          then x : split xs (c - length cs)
          else [Insert $ take c cs]

serverState :: Server o d -> d
serverState (Server start ops) = foldl applyOp start ops

serverStateOp :: Server o d -> o
serverStateOp (Server start ops) = foldl compose (noop start) ops

serverRevision :: Server o d -> Int
serverRevision (Server start ops) = length ops

appendOperation :: Server o d -> o -> Int -> (o,Server o d)
appendOperation (Server start ops) op rev = (op', Server start (ops ++ [op'])) where
  op' = foldl (\a b -> fst $ transform a b) op $ drop rev ops

File: ueb6/client/src/main/scala/OT.scala
import org.scalajs.dom.{document, html, console}
trait Operation {
  type Operation
  type Document

  def noop(doc: Document): Operation
  def compose(a: Operation, b: Operation): Operation
  def transform(a: Operation, b: Operation): (Operation,Operation)  
  def applyOp(doc: Document, op: Operation): Document
  def transformCursor(a: Operation, cursor: Int) : Int
  def reduce(op: Operation) : Operation
}

sealed trait TextAction
case class Retain(n: Int) extends TextAction
case class Insert(cs: String) extends TextAction
case class Delete(n: Int) extends TextAction

object TextOperation extends Operation {
  type Operation = List[TextAction]
  type Document = String

  def reduce(op: Operation) : Operation = op match {
    case (Insert(a)::Insert(b)::as) => reduce(Insert(a+b) :: as)
    case (a :: as) => a :: reduce(as)
    case Nil => Nil
  }

  def retain(n: Int): Operation = List(Retain(n))
  def delete(n: Int): Operation = List(Delete(n))
  def insert(s: String): Operation = List(Insert(s))

  def noop(doc: Document): Operation = List(Retain(doc.length))

  def compose(a: Operation, b: Operation): Operation = (a,b) match {
    case (Nil, Nil)       => Nil
    case (Delete(n)::as, bs) => Delete(n) :: compose(as, bs)
    case (as, Insert(cs)::bs) => 
      Insert(cs) :: compose(as, bs)
    case (Retain(n)::as, Retain(n2)::bs) => 
      if (n > n2) Retain(n2)::compose(Retain(n-n2)::as,bs)
      else if (n < n2) Retain(n)::compose(as,Retain(n2-n)::bs)
      else Retain(n)::compose(as, bs)
    case (Retain(n):: as, Delete(n2)::bs) =>
      if(n > n2) Delete(n2)::compose(Retain(n-n2)::as,bs)
      else if(n < n2) Delete(n)::compose(as,Retain(n2-n)::bs)
      else Delete(n)::compose(as,bs)
    case (Insert(cs)::as,Retain(n)::bs) =>
      if(n > cs.length) Insert(cs)::compose(as,Retain(n-cs.length)::bs)
      else if(n < cs.length) Insert(cs.take(n))::compose(Insert(cs.drop(n))::as, bs)
      else Insert(cs) :: compose(as, bs)
    case (Insert(cs)::as, Delete(n)::bs) =>
      if(n > cs.length) compose(as,Delete(n-cs.length)::bs)
      else if(n < cs.length) compose(Insert(cs.drop(n))::as,bs)
      else compose(as, bs)
  }

  def transform(a: Operation, b: Operation): (Operation,Operation) = (a,b) match {      
    case (Nil,Nil) => (Nil,Nil)
    case (Insert(cs)::as, bs) => 
      val (as_,bs_) = transform(as,bs)
      (Insert(cs)::as_,Retain(cs.length)::bs_)
    case (as, Insert(cs) :: bs) =>
      val (as_,bs_) = transform(as,bs)
      (Retain(cs.length)::as_, Insert(cs)::bs_)
    case (Retain(n):: as, Retain(n2):: bs) =>
      if (n > n2) {
        val (as_,bs_) = transform(Retain(n-n2)::as,bs)
        (Retain(n2)::as_,Retain(n2)::bs_)
      }
      else if (n < n2) {
        val (as_,bs_) = transform(as,Retain(n2-n)::bs)
        (Retain(n)::as_, Retain(n)::bs_)
      }
      else { 
        val (as_,bs_) = transform(as,bs)
        (Retain(n):: as_, Retain(n2):: bs_)
      }
    case (Delete(n)::as, Delete(n2)::bs) =>
      if(n > n2) {
        transform(Delete(n-n2)::as,bs)
      }
      else if(n > n2) {
        transform(as,Delete(n2-n)::bs)
      }
      else {
        transform(as,bs)
      }
    case (Retain(n):: as, Delete(n2)::bs) =>
      if(n > n2) {
        val (as_,bs_) = transform(Retain(n-n2)::as,bs)
        (as_,Delete(n2)::bs_)
      } 
      else if(n < n2) {
        val (as_,bs_) = transform(as,Delete(n2-n)::bs)
        (as_,Delete(n)::bs_)
      }
      else {
        val (as_,bs_) = transform(as,bs)
        (as_, Delete(n)::bs_)
      }
    case (Delete(n):: as, Retain(n2)::bs) =>
      if (n > n2) {
        val (as_,bs_) = transform(Delete(n-n2)::as, bs)
        (Delete(n2)::as_,bs_)
      }
      else if(n < n2) {
        val (as_,bs_) = transform(as,Retain(n2-n)::bs)
        (Delete(n)::as_,bs_)
      }
      else {
        val (as_,bs_) = transform(as,bs)
        (Delete(n)::as_,bs_)
      }
  }

  def transformCursor(op: Operation, cursor: Int) : Int = 
    TextOperation.split(op,cursor).foldRight(cursor) { 
      (o, acc) => o match {
        case Retain(n) => acc
        case Delete(n) => acc - n
        case Insert(cs) => acc + cs.length
      }
    }
  
  def split(op: Operation, c : Int) : Operation = {
    op match {
      case Nil => Nil 
      case (x::xs) => 
        if (c <= 0) Nil 
        else x match {
          case Retain(n) => 
            if (n <= c) x :: split(xs,c-n) 
            else List(Retain(c))
          case Delete(n) =>
            if (n <= c) x :: split(xs,c-n)
            else List(Delete(c))
          case Insert(cs) =>
            if (cs.length <= c) x :: split(xs,c-cs.length) 
            else List(Insert(cs.take(c)))
        }
    }
  }
    def applyOp(doc: Document, op: Operation): Document = (doc,op) match {
    case ("",Nil) => ""
    case (d,Retain(0)::as) => applyOp(d,as)
    case (d,Retain(n)::as) if d.nonEmpty => d.head + applyOp(d.tail,Retain(n-1)::as)
    case (d,Insert(cs)::as) => cs ++ applyOp(d,as)
    case (d,Delete(0)::as) => applyOp(d,as)
    case (d,Delete(n)::as) if d.nonEmpty => applyOp(d.tail, Delete(n-1)::as)
  }
}

trait Client[T <: Operation] { self =>
  val ot: T
  val revision: Int
  val pending: Option[ot.Operation]
  val buffer: Option[ot.Operation]

  private def copy(
    revision: Int = this.revision,
    pending: Option[ot.Operation] = this.pending,
    buffer: Option[ot.Operation] = this.buffer
  ): Client[T] = {
    val (r,p,b) = (revision,pending,buffer)
    new Client[T] {
      val ot = self.ot
      val revision = r
      val pending = p.asInstanceOf[Option[ot.Operation]]
      val buffer = b.asInstanceOf[Option[ot.Operation]]
    }
  }

  def localEdit(op: ot.Operation): (Boolean, Client[T]) = {
    val newPending = pending orElse Some(op)
    val newBuffer = for {
      pending <- pending
    } yield buffer.map(ot.compose(_,op)).map(ot.reduce(_)).getOrElse(op)
    (pending.isEmpty, copy(pending = newPending, buffer = newBuffer))
  }

  def remoteEdit(op: ot.Operation): (ot.Operation, Client[T]) = (pending,buffer) match {
    case (None,None) => (op, copy(revision = revision + 1))
    case (Some(p),None) => 
      val (a,b) = ot.transform (p,op)
      (b, copy(revision = revision + 1, pending = Some(a)))
    case (Some(p),Some(b)) =>
      val (a1,b1) = ot.transform(p,op)
      val (a2,b2) = ot.transform(b,b1)
      (b2, copy(revision + 1, Some(a1), Some(a2)))
  }

  def ack: (Option[ot.Operation], Client[T]) = (buffer, copy(revision + 1, buffer, None))

  def isSynchronized: Boolean = pending.isEmpty

  override def toString = s"Client(\n  revision = $revision,\n  pending = $pending,\n  buffer = $buffer\n)"
}

object Client {
  def empty[T <: Operation](implicit evidence: T): Client[T] = new Client[T] {
    val ot = evidence
    val revision = 0
    val pending = Option.empty[ot.Operation]
    val buffer = Option.empty[ot.Operation]    
  }  
}
